### ISSUE TITLE
ResultServer should use epoll instead

### ISSUE THREAD

**Issue by moonlightelite** [cuckoobox/cuckoo#385]
Hello, I am running cuckoo sandbox with 100+ VMs, and I notice that ResultServer's call to select.select() is occasionally out of file descriptors (The exact error is ValueError: filedescriptor out of range in select() as select() is limited to 1024 FDs (ps. I thought Linux doesn't have this per process limitation?) ).

May I propose using epoll instead? Something like the following. I tested this in my setup and it seems to work, but I am not familiar with the codebase to know all the side effects.

    def wait_sock_or_end(self):
        ep = select.epoll()
        ep.register(self.request, select.EPOLLIN)
        while True:
            if self.end_request.isSet():
                return False
            events = ep.poll(1)
            for fileno, event in events:
                if event & select.EPOLLIN:
                    return True

**Comment by moonlightelite** [comment]
Sorry for the late response.

I don't know if I am using libvirt wrong or what, but it can't really handle 100+ connections reliably. Since Machinery generates a new libvirt connection object per _connect() call, I modified the code to grab conn from a connection object pool instead. This sort of work.

Also the libvirt folks are not very receptive to bug fixes (especially the ESX module)....

**Comment by rep** [comment]
To be honest at this point I don't really see a reason to make improvements on our side to this if there's a simple / easy fix of running several cuckoo instances, covering a maximum number of VMs each. Especially since you mention issues in the underlying libraries as well.

Distributing the load in the architecture design, rather than in our single python process, is actually the right way to go, as there's only so much we could optimize in resultserver / processing / etc.

If you need throughput using Cuckoo we recommend using process_results = off together with "process.py auto" and several cuckoo instances connected to a central database. You can then run more cuckoo processes as the sample volume grows and scale at different locations in the architecture.

Summing up - I don't think we will change anything in the resultserver to make it more efficient for VM counts larger than a few dozen. Running over 100 VMs with one cuckoo process is not recommended and we definitely can't support it given our current development power.

Thus - closing this one. Please feel free to re-open with a PR, or link us a blog post of your setup with suggestions for improvements or something.

**Comment by jbremer** [comment]
Agreed, this is kind of pushing Cuckoo to the limits where that's not really necessary.
For what it's worth, I've had various Cuckoo setups with 20 to 40 VMs ;)
